keps/sig-auth/20190730-oidc-discovery.md

```diff
@@ -0,0 +1,101 @@
+---
+title: Service Account signing key retrieval
+authors:
+  - "@mikedanese"
+  - "@cceckman"
+owning-sig: sig-auth
+participating-sigs:
+  - sig-auth
+reviewers:
+  - "@liggitt"
+  - "@enj"
+  - "@micahhausler"
+  - "@ericchiang"
+approvers:
+  - "@liggitt"
+  - "@enj"
+  - "@micahhausler"
+  - "@ericchiang"
+editor: TBD
+creation-date: 2018-06-26
+last-updated: 2019-07-30
+status: provisional
+replaces:
+  - "https://github.com/kubernetes/community/pull/2314/"
+---
+
+# Service Account signing key retrieval
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+A non-resource API will be added to the apiserver that will expose an [OIDC
+discovery](https://openid.net/specs/openid-connect-discovery-1_0.html)
+compatible configuration. This will allow clients to retrieve a list of
+currently valid signing keys as an array of JWKs. For example, if the service
+account issuer is configured with the name `https://dev.cluster.internal`, the
+apiserver could expose the following configuration:
+
+ ```
+> GET /.well-known/openid-configuration
```

Comment:
What default roles will have permission on these endoints?

Reply:
Documented inline. Empirically tested on a GKE cluster:

```
$ URL=$(kubectl config view --minify -o json | jq -r '.clusters[] | .cluster.server')
$ alias gcurl='curl -H "Authorization: Bearer $(gcloud auth print-access-token)" -H "Content-Type: application/json"'
$ gcurl -sk "$URL/apis/rbac.authorization.k8s.io/v1/clusterroles" | jq '.items[] | select(any(.rules[]; has("nonResourceURLs")))' 
```

`cluster-admin` and `system:controller:clusterrole-aggregation-controller` have access to all `nonResourceURLs`, and would therefore have access to these paths.

`system:discovery` has access to *some* `nonResourceURLs`, but not all - in particular, not to `/.well-known` or '/serviceaccountkeys`.


We could propose changing default roles (e.g. `system:discovery`, `view') to grant access to these paths.

Reply:
Could we make this truly anonymous - i.e. no authn / authz, a separate minimal http handler?  Then we could also put a health check endpoint there and finally solve the anonymous-access-is-required-but-generally-scary problem!

Reply:
Unrelated to this KEP, but we also have a need for a secondary port/listener like this for `/healthz` that is not covered by authn/authz.

Reply:
That's what `system:public-info-viewer` is for. I don't think we should special case this handler.

Reply:
Updated to say the `system:public-info-viewer` role will be granted access.

Reply:
Does `system:public-info-viewer` require you to be authenticated with the cluster before reading the OIDC endpoints? Does this mean federating KSAs will always require a proxy to serve unauthenticated endpoints?

Reply:
default RBAC policy grants that role to all viewers (authenticated and unauthenticated): https://github.com/kubernetes/kubernetes/blob/release-1.16/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-role-bindings.yaml#L135-L154

clusters may opt to disable anonymous requests entirely with `--anonymous-auth=false`

Reply:
Ok, that seems like a fair position for now.